okhttp-urlconnection/src/test/java/okhttp3/JavaNetAuthenticatorTest.java

```diff
@@ -0,0 +1,177 @@
+package okhttp3;
+
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+import org.junit.runners.Parameterized.Parameter;
+import org.junit.runners.Parameterized.Parameters;
+
+import javax.net.SocketFactory;
+import java.net.PasswordAuthentication;
+import java.net.Proxy;
+import java.net.ProxySelector;
+import java.net.UnknownHostException;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+
+import static java.net.HttpURLConnection.HTTP_PROXY_AUTH;
+import static java.net.HttpURLConnection.HTTP_UNAUTHORIZED;
+import static java.net.InetSocketAddress.createUnresolved;
+import static java.net.Proxy.Type.HTTP;
+import static java.util.Arrays.asList;
+import static java.util.Collections.singletonList;
+import static java.util.UUID.randomUUID;
+import static okhttp3.Credentials.basic;
+import static okhttp3.Protocol.HTTP_1_1;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertNull;
+
+@RunWith(Parameterized.class)
+public class JavaNetAuthenticatorTest {
+  private static JavaNetAuthenticator javaNetAuthenticator = new JavaNetAuthenticator();
+
+  @Parameter public List<String> challenges;
+  @Parameter(1) public String realmToAnswer;
+  @Parameter(2) public boolean proxyAuthenticate;
+
+  private int responseCode;
+  private String authenticateHeaderName;
+  private String authorizationHeaderName;
+  private String username = randomUUID().toString();
+  private String password = randomUUID().toString();
```

Comment:
Surely I could do that, but are you really sure?
Imho randomness and tests are best friends in the right situations.
It is not so much the randomness that I wanted here, but the uniqueness.

How it is right now, each test method has its own set of credentials, so it is made sure that one test does not pass because some previous test, maybe even in another class, did set an authenticator and didn't properly unset it or something similar.

Alternatively I could maybe have a static counter variable that increases on each test run so that you have `username1`, `username2` and so on if you would like that better.

Reply:
Wanting uniqueness between tests is surprising. My preference is for every unit test to cover it's own subject.

If there's a risk that tests cross-contaminate we should test that independently. (There isn't as far as I can tell, however, and we don't need useless tests.)

Reply:
The tests are not useless, but tried to be more robust, but ok, I hard-coded "username" and "password" now.
You might want to have a look at `okhttp3.RequestTest` though, it also uses `java.util.UUID.randomUUID()` multiple times.

Reply:
Ah, I see. Yes, I agree that these tests would be strange.